### PR TITLE
Add quiescence

### PR DESCRIPTION
**Pull request by remyers**
Porting quiescence PR from Eclair.

### CODE REVIEW

**Comment by remyers** [comment]
This is still in progress while I work out some issues with the last two tests related to what happens when an incoming htlc is about to timeout while quiescent.

**Comment by remyers** [comment]
> This is still in progress while I work out some issues with the last two tests related to what happens when an incoming htlc is about to timeout while quiescent.

I think this is fixed with [c5140e6](https://github.com/ACINQ/lightning-kmp/pull/558/commits/c5140e6ba8830df44dcd1d8cd1f2e398a29051f0). But it could be that I'm missing some subtly here so worth discussing separately. 

In particular, I do not see why `almostTimedOutIncomingHtlcs` was ignoring expiring htlcs if there was no pending local `UpdateFulfillHtlc`.  If we are quiescent we will not have processed the `UpdateFulfillHtlc` yet so that filter prevents the channel from being closed during quiescence.

**Comment by remyers** [comment]
> > This is still in progress while I work out some issues with the last two tests related to what happens when an incoming htlc is about to timeout while quiescent.
> 
> I think this is fixed with [c5140e6](https://github.com/ACINQ/lightning-kmp/pull/558/commits/c5140e6ba8830df44dcd1d8cd1f2e398a29051f0). But it could be that I'm missing some subtly here so worth discussing separately.
> 
> In particular, I do not see why `almostTimedOutIncomingHtlcs` was ignoring expiring htlcs if there was no pending local `UpdateFulfillHtlc`. If we are quiescent we will not have processed the `UpdateFulfillHtlc` yet so that filter prevents the channel from being closed during quiescence.

I changed `almostTimedOutIncomingHtlcs` to not check for `UpdateFulfillHtlc` but looking at Eclair I think this is a mistake. Instead I should handle it more like [`handleNewBlock`](https://github.com/ACINQ/eclair/blob/35e318e951f60777ffcd8efeddab8faab80bfa4e/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala#L2657C11-L2657C11). Instead of checking the `pendingCommands` db I should check if there are any matching `UpdateFulfillHtlc` for expiring incoming htlcs in the `settlementStash`.

**Comment by remyers** [comment]
> This is still in progress while I work out some issues with the last two tests related to what happens when an incoming htlc is about to timeout while quiescent.

I believe the correct solution for htlcs that timeout is that on disconnect/reconnect or when quiescence ends, incoming htlcs are processed by the peer. If they are close to timing out, they will be failed at that point.

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt [hunk 1]
```diff
@@ -345,9 +359,99 @@ data class Normal(
...
+                                            logger.info { "initiating splice with local.amount=${spliceInit.fundingContribution} local.push=${spliceInit.pushAmount}" }
+                                            Pair(this@Normal.copy(spliceStatus = SpliceStatus.Requested(spliceStatus.command, spliceInit)), listOf(ChannelAction.Message.Send(spliceInit)))
+                                        }
+                                    } else {
+                                        logger.warning { "cannot initiate splice, channel not quiescent" }
+                                        spliceStatus.command.replyTo.complete(ChannelCommand.Commitment.Splice.Response.Failure.ChannelNotQuiescent)
+                                        val actions = buildList {
+                                            add(ChannelAction.Message.Send(Warning(channelId, InvalidSpliceNotQuiescent(channelId).message)))
+                                            add(ChannelAction.Disconnect)
+                                            // pending settlement commands will be replayed on reconnect
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt:428]: nit: remove (and everywhere else)

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt [hunk 2]
```diff
@@ -345,9 +359,99 @@ data class Normal(
...
+                                    if (commitments.isQuiescent()) {
+                                        val parentCommitment = commitments.active.first()
+                                        val fundingContribution = FundingContributions.computeSpliceContribution(
+                                            isInitiator = true,
+                                            commitment = parentCommitment,
+                                            walletInputs = spliceStatus.command.spliceIn?.walletInputs ?: emptyList(),
+                                            localOutputs = spliceStatus.command.spliceOutputs,
+                                            targetFeerate = spliceStatus.command.feerate
+                                        )
+                                        val commitTxFees = Transactions.commitTxFee(commitments.params.remoteParams.dustLimit, parentCommitment.remoteCommit.spec)
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt:391]: Actually that is incorrect, we should set this to 0 if we're not the channel initiator (and thus don't pay the commit tx fee). This is a (small) bug on eclair, where we should only take this into account when we're the channel initiator: can you open a PR to fix this on eclair?

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt [hunk 3]
```diff
@@ -380,8 +484,8 @@ data class Normal(
                                 val nextState = this@Normal.copy(spliceStatus = SpliceStatus.InProgress(replyTo = null, session, localPushAmount = 0.msat, remotePushAmount = cmd.message.pushAmount, origins = cmd.message.origins))
                                 Pair(nextState, listOf(ChannelAction.Message.Send(spliceAck)))
                             } else {
-                                logger.info { "rejecting splice attempt: channel is not idle" }
-                                Pair(this@Normal.copy(spliceStatus = SpliceStatus.Aborted), listOf(ChannelAction.Message.Send(TxAbort(channelId, InvalidSpliceChannelNotIdle(channelId).message))))
+                                logger.warning { "cannot initiate splice, channel is not quiescent" }
+                                Pair(this@Normal, listOf(ChannelAction.Message.Send(Warning(channelId, InvalidSpliceNotQuiescent(channelId).message))))
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt:488]: Why did you change that to a `Warning`? I think this really should be `TxAbort`, otherwise the other node will be stuck expecting our `splice_ack`?

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt [hunk 4]
```diff
@@ -380,8 +484,8 @@ data class Normal(
                                 val nextState = this@Normal.copy(spliceStatus = SpliceStatus.InProgress(replyTo = null, session, localPushAmount = 0.msat, remotePushAmount = cmd.message.pushAmount, origins = cmd.message.origins))
                                 Pair(nextState, listOf(ChannelAction.Message.Send(spliceAck)))
                             } else {
-                                logger.info { "rejecting splice attempt: channel is not idle" }
-                                Pair(this@Normal.copy(spliceStatus = SpliceStatus.Aborted), listOf(ChannelAction.Message.Send(TxAbort(channelId, InvalidSpliceChannelNotIdle(channelId).message))))
+                                logger.warning { "cannot initiate splice, channel is not quiescent" }
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt:487]:
```suggestion
                                logger.warning { "rejecting splice attempt: channel is not quiescent" }
```

**Review by t-bast — COMMENTED** [review]
I think the HTLC reprocessing is actually simpler than what you thought, see one of my comments.

File: src/commonMain/kotlin/fr/acinq/lightning/Features.kt [hunk 5]
```diff
@@ -242,6 +242,13 @@ sealed class Feature {
         override val scopes: Set<FeatureScope> get() = setOf(FeatureScope.Init)
     }
 
+    @Serializable
+    object Quiescence : Feature() {
+        override val rfcName get() = "option_quiescence"
+        override val mandatory get() = 34
+        override val scopes: Set<FeatureScope> get() = setOf(FeatureScope.Init)
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/Features.kt]: Fixed in [982d5d2](https://github.com/ACINQ/lightning-kmp/pull/558/commits/982d5d22fa484ef7be1bb7925664d50f32a4ca4b).

File: src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt [hunk 6]
```diff
@@ -287,9 +287,10 @@ class Peer(
         }
     }
 
-    fun disconnect() {
+    suspend fun disconnect() {
         if (this::socket.isInitialized) socket.close()
         _connectionState.value = Connection.CLOSED(null)
+        input.send(Disconnected)
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt:293]: I was thinking of the situation where you might (still) have more than one channel. This is the only place where you actually choose to disconnect so I thought it would be necessary to make sure the channels were left in the 'Offline' state. 

However, if the working assumption is that the state doesn't change on disconnect, then I understand the comment in`establishConnection`.

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt (hunk 4 above)
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt]: fixed in [982d5d2](https://github.com/ACINQ/lightning-kmp/pull/558/commits/982d5d22fa484ef7be1bb7925664d50f32a4ca4b)

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt (hunk 3 above)
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt]: This was a mistake - good catch! fixed in [982d5d2](https://github.com/ACINQ/lightning-kmp/pull/558/commits/982d5d22fa484ef7be1bb7925664d50f32a4ca4b).

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt (hunk 2 above)
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt]: fixed in [982d5d2](https://github.com/ACINQ/lightning-kmp/pull/558/commits/982d5d22fa484ef7be1bb7925664d50f32a4ca4b) and I'll also open a PR in scala with the same fix.

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt (hunk 1 above)
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt]: fixed in [982d5d2](https://github.com/ACINQ/lightning-kmp/pull/558/commits/982d5d22fa484ef7be1bb7925664d50f32a4ca4b).

File: src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt [hunk 7]
```diff
@@ -215,7 +214,22 @@ data class Normal(
                                         if (result.value.first.changes.localHasChanges()) {
                                             actions.add(ChannelAction.Message.SendToSelf(ChannelCommand.Commitment.Sign))
                                         }
-                                        Pair(nextState, actions)
+                                        // If we're now quiescent, we may send our stfu message.
+                                        if (result.value.first.localIsQuiescent()) {
+                                            when (spliceStatus) {
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/channel/states/Normal.kt]: I like this version better too. Fixed in [982d5d2](https://github.com/ACINQ/lightning-kmp/pull/558/commits/982d5d22fa484ef7be1bb7925664d50f32a4ca4b).

File: src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt:293]: > I was thinking of the situation where you might (still) have more than one channel.

I don't understand why it would make anything different? If you disconnect, that disconnects all channels?

The reason we process this in `establishConnection` is because there could otherwise be a race condition, where `Disconnect` ends up being processed *after* the reconnection, which completely breaks the state machine. This happens somewhat frequently because when we're disconnected, we immediately try reconnecting.

Also, you may not be aware that you've been disconnected (if you're intermittently running in the background for example). That's ok, the only important thing to ensure is that you clean up state from a previous connection before reconnecting, which is what we currently do.

[the remark this answers is not in the document]
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt]: Reverted back in [990c98e](https://github.com/ACINQ/lightning-kmp/pull/558/commits/990c98eebd8e91548ebe5493583fb8374d794e4c). I was missing the fact that  ultimately the shared connection state is held by the `socket` object in `Peer` which applies to all channels and that the `Disconnected` message is used as a prelude to a re-connection.

File: src/commonMain/kotlin/fr/acinq/lightning/wire/LightningMessages.kt [hunk 8]
```diff
@@ -853,6 +853,28 @@ data class ChannelReady(
...
+        LightningCodecs.writeBytes(channelId, out)
+        LightningCodecs.writeByte(if (initiator) 1 else 0, out)
+    }
+    companion object : LightningMessageReader<Stfu> {
+        const val type: Long = 2
+
+        override fun read(input: Input): Stfu {
+            return Stfu(
+                ByteVector32(LightningCodecs.bytes(input, 32)),
+                LightningCodecs.byte(input).toByte() == 1.toByte()
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/wire/LightningMessages.kt]: fixed in [0bb438e](https://github.com/ACINQ/lightning-kmp/pull/558/commits/0bb438ee40170e698eb8e1610c651aa660f625e2).

File: src/commonMain/kotlin/fr/acinq/lightning/channel/Commitments.kt [hunk 9]
```diff
@@ -554,15 +552,62 @@ data class Commitments(
...
+    // They signed it first, so the HTLC will first appear in our commitment tx, and later on in their commitment when we subsequently sign it.
+    // That's why we need to look in *their* commitment with direction=OUT.
+    //
+    // We also need to filter out htlcs that we already settled and signed (the settlement messages are being retransmitted).
+    fun reprocessIncomingHtlcs(log: MDCLogger? = null): List<ChannelAction> {
+        val alreadySettled = changes.localChanges.signed.filterIsInstance<HtlcSettlementMessage>().map { it.id }.toSet()
+        val htlcsToReprocess = latest.remoteCommit.spec.htlcs.outgoings().filter { !alreadySettled.contains(it.id) }
+        log?.debug { "re-processing signed IN: $htlcsToReprocess" }
+
+        return processIncomingHtlcs(htlcsToReprocess)
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/channel/Commitments.kt]: Good suggestions! I've reverted the changes that added `processIncomingHtlcs` and just kept the simplified `reprocessIncomingHtlcs`. Fixed in [0e3d3d7](https://github.com/ACINQ/lightning-kmp/pull/558/commits/0e3d3d71da4923ca69ed5e47087fbd9816f717c8).

File: src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt [hunk 10]
```diff
@@ -701,6 +702,8 @@ class Peer(
                     }
 
                     is ChannelAction.EmitEvent -> nodeParams._nodeEvents.emit(action.event)
+
+                    is ChannelAction.Disconnect -> disconnect()
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt]: Fixed in [990c98e](https://github.com/ACINQ/lightning-kmp/pull/558/commits/990c98eebd8e91548ebe5493583fb8374d794e4c)

**Comment by remyers** [comment]
I noticed while working on the test `receive SpliceInit when channel is not quiescent` that currently our peer can send `SpliceInit` immediately without negotiating quiescence first and if the peers are idle, it will not reject it. I can't think of a reason to reject `SpliceInit` in this situation except that it means something is wrong with your peer.

**Comment by remyers** [comment]
> I noticed while working on the test `receive SpliceInit when channel is not quiescent` that currently our peer can send `SpliceInit` immediately without negotiating quiescence first and if the peers are idle, it will not reject it. I can't think of a reason to reject `SpliceInit` in this situation except that it means something is wrong with your peer.

Fixed in [ef52264](https://github.com/ACINQ/lightning-kmp/pull/558/commits/ef52264a36ecf8388194e7c217e9dc5b046a613f). I'll also make this change in the Eclair PR with feedback from this PR.

**Comment by remyers** [comment]
Is it correct that if our pending outoing htlc times out, we should force close during a splice? This is new situation we didn't have to consider before quiescence. It seems like the risk is on the htlc receiver's side so we shouldn't need to do a preemptive force close. We might have an unprocessed htlc success or fail that came in while we were quiescent. Since our node must be the source of the htlc, we don't risk our upstream force closing on us.

**Comment by t-bast** [comment]
> It seems like the risk is on the htlc receiver's side so we shouldn't need to do a preemptive force close.

That's a very good point, in theory we wouldn't need to force-close, even if the HTLC has expired, because we have no funds at risk.

There are two different cases here:

* Our peer has the preimage for that HTLC
* Our peer doesn't have the preimage for that HTLC (it was failed on their side or they forced close the downstream channel)

In the first case, our peer will force-close before the HTLC times out anyway because they have funds at risk.

But in the second case, our peer won't preemptively force-close (at least eclair won't). So it could be beneficial if we don't force-close either, because we can wait any amount of time before receiving `update_fail_htlc`. But that opens the door to funds being stuck: if you never force-close, your peer can keep timed out HTLCs forever in the commitment, which locks some of your liquidity. Not force-closing would increase the trust required in your peer's behavior. This shouldn't happen anyway, since splice negotiation should always be quick, so I think we should keep the default behavior of force-closing when HTLCs time out.

**Review by t-bast — COMMENTED** [review]
I've spent some time on the quiescence tests, reworked them and made some refactoring in https://github.com/ACINQ/lightning-kmp/commit/3ca8b0654c3614521e38bfc0bc12451c04688ecc, could you include this commit?

It made me discover that we aren't correctly handling concurrency between `stfu` and `shutdown`. This commit contains fixes for all my comments below.

Can you also rebase to fix the conflict with `Peer.kt`?

File: src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+            assertIs<LNChannel<Normal>>(sender2)
+            assertIs<LNChannel<Normal>>(receiver1)
+            return Triple(sender2, receiver1, spliceInit)
+        }
+        assertIs<LNChannel<Normal>>(sender1)
+        assertIs<LNChannel<Normal>>(receiver)
+        return Triple(sender1, receiver, stfu)
+    }
+
+    private fun receiveSettlementCommand(alice: LNChannel<Normal>, bob: LNChannel<Normal>, sendInitialStfu: Boolean, resetConnection: Boolean = false) {
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt]: I'll try to keep that in mind; I thought if I made my tests for lightning-kmp match closely their equivalents in Eclair would help with review. I should have made a few more passes after the equivalent tests were working to clean it up.

File: src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+        assertEquals(actionsBob3.size, 0)
+    }
+
+    @Test
+    fun `recv Shutdown message before initiator receives stfu from remote` () {
+        val (alice, bob) = init()
+        val (alice1, _, _) = initiateQuiescence(alice, bob, sendInitialStfu = false)
+        val forbiddenMsg = Shutdown(alice.channelId, bob.commitments.params.localParams.defaultFinalScriptPubKey)
+        // handle Shutdown normally
+        alice1.process(ChannelCommand.MessageReceived(forbiddenMsg)).second.findOutgoingMessage<Shutdown>()
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt]: You're right - it was also not handled in Eclair. I'll add that to my PR for back-ported changes from lightning-kmp.

File: src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt [hunk 13]
```diff
@@ -0,0 +1,551 @@
...
+            actionsAlice3.findOutgoingMessage<Warning>()
+            actionsAlice3.has<ChannelAction.Disconnect>()
+            val (_, actionsBob3) = bob3.process(ChannelCommand.MessageReceived(it))
+            actionsBob3.findOutgoingMessage<Warning>()
+            actionsBob3.has<ChannelAction.Disconnect>()
+        }
+    }
+
+    @Test
+    fun `fulfilled htlc times out while offline after quiescence` () {
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt]: Good catch - this test doesn't make any sense.

File: src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+        runBlocking {
+            withTimeout(100) {
+                assertIs<ChannelCommand.Commitment.Splice.Response.Failure.ConcurrentRemoteSplice>(cmdAlice.replyTo.await())
+            }
+        }
+        actionsBob4.findOutgoingMessage<SpliceInit>()
+    }
+
+    @Test
+    fun `initiate quiescence concurrently with pending changes on non-initiator side` () {
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt]: They are slightly different. In both tests, alice  and bob both initiate a splice.

File: src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt [hunk 15]
```diff
@@ -0,0 +1,551 @@
...
+        assertEquals(1, htlcTimeoutTxs.size)
+        val htlcSuccessTxs = lcp.htlcSuccessTxs()
+        assertEquals(0, htlcSuccessTxs.size)
+
+        // Valid txs should be detected:
+        htlcTimeoutTxs.forEach { tx -> assertTrue(lcp.isHtlcTimeout(tx.tx)) }
+    }
+
+    @Test
+    fun `incoming htlc timeout during quiescence negotiation with pending htlc failure or success` () {
```
[the remark this answers is not in the document]
Comment (remyers) [reply src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt]: In know this situation isn't possible for lightning-kmp, but would Alice be at risk if she had routed the htlc and received the preimage from an upstream peer? That's how the test goes for Eclair.

[the remark this answers is not in the document]
Comment (t-bast) [reply src/commonTest/kotlin/fr/acinq/lightning/channel/states/QuiescenceTestsCommon.kt]: Yes exactly: if we supported routing payments, then the behavior would be different. So eclair behaves differently from `lightning-kmp` in that scenario.

**Review by t-bast — COMMENTED** [review]
LGTM! Can you squash that into a single commit? It will make it easier for future rebase while you work on the follow-up PR that handles splicing with pending HTLCs.

**Review by t-bast — COMMENTED** [review]
LGTM after rebase, still waiting to integrate along with the splice-htlc follow-up PR.

**Review by t-bast — COMMENTED** [review]
ACK https://github.com/ACINQ/lightning-kmp/pull/558/commits/32d4f29779c84fbc7c274a69907ea2adc2b741a4

**Comment by t-bast** [comment]
Included as part of #568
